GHC 9.2 support

bergmark opened the pull request:
I somehow missed #23 until i was done :-h But that one doesn't work on older GHCs.

athas commented:
Is this amount of changes needed?  I fixed it (I think) by adding two more ifdefs directly in the code, which seems to fit how haskell-src-meta is currently implemented anyway.

```
diff --git a/haskell-src-meta/haskell-src-meta.cabal b/haskell-src-meta/haskell-src-meta.cabal
index dbfe318..aa97784 100644
--- a/haskell-src-meta/haskell-src-meta.cabal
+++ b/haskell-src-meta/haskell-src-meta.cabal
@@ -22,7 +22,7 @@ library
                    haskell-src-exts >= 1.18 && < 1.24,
                    pretty >= 1.0 && < 1.2,
                    syb >= 0.1 && < 0.8,
-                   template-haskell >= 2.10 && < 2.18,
+                   template-haskell >= 2.10 && < 2.19,
                    th-orphans >= 0.12 && < 0.14

   if impl(ghc < 7.8)
diff --git a/haskell-src-meta/src/Language/Haskell/Meta/Syntax/Translate.hs b/haskell-src-meta/src/Language/Haskell/Meta/Syntax/Translate.hs
index 6b795bf..6dca9e1 100644
--- a/haskell-src-meta/src/Language/Haskell/Meta/Syntax/Translate.hs
+++ b/haskell-src-meta/src/Language/Haskell/Meta/Syntax/Translate.hs
@@ -222,7 +222,11 @@ instance ToPat (Exts.Pat l) where
     TH.DoublePrimL r'' -> TH.DoublePrimL (negate r'')
     _                  -> nonsense "toPat" "negating wrong kind of literal" l
   toPat (Exts.PInfixApp _ p n q) = TH.UInfixP (toPat p) (toName n) (toPat q)
+#if MIN_VERSION_template_haskell(2,18,0)
+  toPat (Exts.PApp _ n ps) = TH.ConP (toName n) [] (fmap toPat ps)
+#else
   toPat (Exts.PApp _ n ps) = TH.ConP (toName n) (fmap toPat ps)
+#endif
   toPat (Exts.PTuple _ Exts.Boxed ps) = TH.TupP (fmap toPat ps)
   toPat (Exts.PTuple _ Exts.Unboxed ps) = TH.UnboxedTupP (fmap toPat ps)
   toPat (Exts.PList _ ps) = TH.ListP (fmap toPat ps)
diff --git a/haskell-src-meta/src/Language/Haskell/Meta/Utils.hs b/haskell-src-meta/src/Language/Haskell/Meta/Utils.hs
index 6cdadcb..51f4eac 100644
--- a/haskell-src-meta/src/Language/Haskell/Meta/Utils.hs
+++ b/haskell-src-meta/src/Language/Haskell/Meta/Utils.hs
@@ -333,7 +333,11 @@ fromDataConI (DataConI dConN ty _tyConN) =
   let n = arityT ty
   in replicateM n (newName "a")
       >>= \ns -> return (Just (LamE
+#if MIN_VERSION_template_haskell(2,18,0)
+                    [ConP dConN [] (fmap VarP ns)]
+#else
                     [ConP dConN (fmap VarP ns)]
+#endif
 #if MIN_VERSION_template_haskell(2,16,0)
                     (TupE $ fmap (Just . VarE) ns)
 #else
```

I have no particular opinion on adding a compatibility module, but for the short-term goal of ensuring GHC 9.2 support, it does not seem needed.  Can you weigh in @DanBurton?

thomasjm commented:
Looks like most of this PR just consolidates a bunch of existing ifdefs into a compat module, which seems like a good/innocuous change to me...

Ericson2314 commented:
I agree this consolidation is good. You might be able to get some of this stuff from `th-abstraction`, too.

bergmark commented:
I changed CI to github and updated the stack.yamls as well.

@DanBurton would you like help maintaining this library? I'm not a direct user but often run across it transitively.

athas commented:
I am in the same position, and haskell-src-meta seems to have a long tradition of passing to new maintainers!

bergmark commented:
CI failed @ https://github.com/bergmark/haskell-src-meta/actions/runs/1700083245

It looks like this is known https://github.com/DanBurton/haskell-src-meta/issues/20. Perhaps we should just drop GHC < 8.2 support?

DanBurton commented:
> @DanBurton would you like help maintaining this library? I'm not a direct user but often run across it transitively.

Yes, @bergmark, in fact I'm happy to hand it over to you if you'd like. Or I can just give you maintainer permissions on this repo & on hackage if you'd prefer.

bergmark commented:
Great! To get around the issue of only being able to have one owner on personal repositories, please transfer it to the new haskell-party organization (you should be able to do that after accepting the invite).

DanBurton commented:
Thanks @bergmark, I've performed the repo transfer.
